rinex-cli/src/graph/record/navigation.rs

```diff
@@ -3,70 +3,163 @@ use plotly::common::{Mode, Visible};
 use rinex::navigation::Ephemeris;
 use rinex::prelude::*;
 
+use itertools::Itertools;
+use std::collections::{BTreeMap, HashMap};
+
+use sp3::SP3;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum ProductType {
+    Radio,
+    HighPrecisionSp3,
+    HighPrecisionClk,
+}
+
+/// Clock states from all products provided by User
+type CtxClockStates = HashMap<ProductType, BTreeMap<SV, Vec<(Epoch, (f64, f64, f64))>>>;
+/// Clock Corrections from all products provided by User
+type CtxClockCorrections = HashMap<ProductType, BTreeMap<SV, Vec<(Epoch, Duration)>>>;
+
 pub fn plot_sv_nav_clock(ctx: &RnxContext, plot_ctx: &mut PlotContext) {
-    let mut clock_plot_created = false;
     if let Some(nav) = ctx.nav_data() {
-        /*
-         * Plot SV Clock Offset/Drift
-         * one plot (2 Y axes) for both Clock biases
-         * and clock drift
-         */
-        for (sv_index, sv) in nav.sv().enumerate() {
-            if sv_index == 0 {
-                plot_ctx.add_timedomain_2y_plot(
-                    "SV Clock Bias",
-                    "Clock Bias [s]",
-                    "Clock Drift [s/s]",
-                );
-                trace!("sv clock plot");
-                clock_plot_created = true;
-            }
-            let sv_epochs: Vec<_> = nav
-                .sv_clock()
-                .filter_map(
-                    |(epoch, svnn, (_, _, _))| {
-                        if svnn == sv {
-                            Some(epoch)
+        let nav_sv = nav.sv().collect::<Vec<_>>();
+        let clk = ctx.clk_data();
+        let sp3 = ctx.sp3_data();
+        let clock_states = ctx_sv_clock_states(nav, &nav_sv, clk, sp3);
+        plot_sv_clock_states(&clock_states, &nav_sv, plot_ctx);
+
+        if let Some(obs) = ctx.obs_data() {
+            let clock_corrections = ctx_sv_clock_corrections(obs, nav, clk, sp3);
+            plot_sv_clock_corrections(&clock_corrections, plot_ctx);
+            plot_system_time(&clock_states, &clock_corrections, &nav_sv, plot_ctx);
+        } else {
+            info!("adding OBS RINEX will provide clock corrections graphs");
+        }
+    } else {
+        warn!("missing navigation (orbits) data");
+    }
+}
+
+/*
+ * Determines all clock states
+ */
+fn ctx_sv_clock_states(
+    nav: &Rinex,
+    nav_sv: &Vec<SV>,
+    clk: Option<&Rinex>,
+    sp3: Option<&SP3>,
+) -> CtxClockStates {
+    let mut states = CtxClockStates::new();
+    for product in [
+        ProductType::Radio,
+        ProductType::HighPrecisionSp3,
+        ProductType::HighPrecisionClk,
+    ] {
+        match product {
+            ProductType::Radio => {
+                let mut tree = BTreeMap::<SV, Vec<(Epoch, (f64, f64, f64))>>::new();
+                for (t, sv, (bias, drift, driftr)) in nav.sv_clock() {
+                    if let Some(inner) = tree.get_mut(&sv) {
```

Comment:
Nit: you can do something like `tree.entry(sv).or_default().push((t, (bias, drift, driftr)))`.

Reply:
pretty neat, I should become more familiar with the `.entry()` API